Fix prompt-setting deprecation

h35514n opened the pull request:
As of Pry v0.13.0, a [deprecation](https://github.com/pry/pry/blob/master/CHANGELOG.md#deprecations) warning is issued due to the way jazz_fingers sets the Pry prompt (see: https://github.com/pry/pry/pull/1877 and #22).

The change in 594223d appears to be all that's required to fix the deprecation minimally.

Commit beaadf4 refactors to use the approach suggested by Pry. The main benefit of this change is that adding the jazz fingers prompt to the `Pry::Prompt` hash allows for switching the
prompt dynamically from the Pry command line using `change-prompt jazz_fingers`.

<details>
<summary>changing the prompt with <code>change-prompt</code></summary>

```rb
2.7.0 (jazz_fingers)[1] » change-prompt --list
# default
# The default Pry prompt. Includes information about the current expression
# number, evaluation context, and nesting level, plus a reminder that you're
# using Pry.
# 
# simple
# A simple `>>`.
# 
# nav
# A prompt that displays the binding stack as a path and includes information
# about _in_ and _out_.
# 
# shell
# A prompt that displays `$PWD` as you change it.
# 
# none
# Wave goodbye to the Pry prompt.
# 
# jazz_fingers (selected)
# JazzFingers prompt

2.7.0 (jazz_fingers)[2] » change-prompt default

[3] pry(main)> change-prompt simple

>> change-prompt jazz_fingers

2.7.0 (jazz_fingers)[5] »
```
</details>

The convenience afforded by `Pry::Prompt.add` also made it easy to add support for displaying the local context (examples below), so that's included in beaadf4 as well.

```rb
2.5.5 (jazz_fingers)[1] » cd Class1::Class2::Class3::Class4
2.5.5 (Class1::Class2::Class3::Class4)[2] »
```

Commit 2453e7e abbreviates object paths displayed in object instances, which can get very long  in the context of test runs:

```
    1559:         require "pry"; binding.pry
 => 1560:         bike.set_location_info
    1561: 

2.5.5 (#<RSpec::ExampleGroups::Bike::SetLocationInfo::GivenNoCreationOrgLocation>)[1] » 
```

The abbreviation is performed by taking as much of the tail of the object path entries list as possible, subject to a max character count constraint. 

An example (Given a `binding.pry` in `Class10#some_method`):

```rb
2.5.5 (jazz_fingers)[4] » Class1::Class2::Class3::Class4::Class5::Class6::Class7::Class8::Class9::Class10.new.some_method
2.5.5 (#<...Class7::Class8::Class9::Class10>)[1] » exit
```

Alternatively, another way to abbreviate the context string for object instances is to only include the root and leaf nodes in the object path (this isn't currently implemented, but can be added with a couple of lines):

```rb
2.7.0 (#<Class1::...::Class10>)[1] » 
```
Happy to modify any of the above as desired.

Fixes #22

Commit:
Fix deprecation, update required Pry version

Commit:
Refactor to use Pry::Prompt.add

Adding the jazz fingers prompt to the Pry::Prompt hash allows for
switching the prompt dynamically using `change-prompt jazz_fingers`.

This patch adds support for displaying the local context (e.g.,
`ModuleName::ClassName`) when `cd`-ing into an object.

h35514n commented on lib/jazz_fingers/prompt.rb:
```diff
@@ -32,44 +34,55 @@ def separator
     end
 
-    def text(object, level)
-      level = 0 if level < 0
```
This method (`#text`) doesn't appear to be used anywhere, so I removed it — let me know if I should revert this change.

Commit:
Abbreviate context string so prompt doesn't overflow

When `cd`ing into an object or using `binding.pry` to trigger a
debugging session, the context string can overflow the prompt.

Example:

```
2.7.0 (jazz_fingers)[4] » Class1::Class2::Class3::Class4::Class5::Class6::Class7::Class8::Class9::Class10.new.speak
2.7.0 (#<Class1::Class2::Class3::Class4::Class5::Class6::Class7::Class8::Class9::Class10>)[1] »
```

This patch abbreviates the context string to display only its tail.

Result:

```
2.7.0 (jazz_fingers)[3] » Class1::Class2::Class3::Class4::Class5::Class6::Class7::Class8::Class9::Class10.new.speak
2.7.0 (#<...::Class7::Class8::Class9::Class10>)[1] »
```

Commit:
Tweak Pry::Prompt description

plribeiro3000 commented on CHANGELOG.md:
```diff
@@ -1,5 +1,10 @@
+* Use non-deprecated method of setting prompt (see [Pry PR #1877][pry-1877])
+
+[pry-1877]: https://github.com/pry/pry/pull/1877
+
```
Could you leave this to be done in another commit/PR please?

h35514n replied:
No prob - updated in b54e2e6

plribeiro3000 commented on jazz_fingers.gemspec:
```diff
@@ -21,7 +21,7 @@ Gem::Specification.new do |gem|
   gem.required_ruby_version = '>= 2.0'
   gem.add_runtime_dependency 'awesome_print', '~> 1.6'
-  gem.add_runtime_dependency 'pry', '~> 0.10'
+  gem.add_runtime_dependency 'pry', '>= 0.13.0'
```
Hmmm. Does this mean that someone that can't use `pry >= 0.13.0` will have to keep using old version of `jazz_fingers`?

Maybe we can loose this a bit and work on the code to work on both constraints?

h35514n replied:
Makes sense. I'm taking a look at that now. Seems like the current Pry dependency requirement might be incorrect? I'm trying to find the most recent Pry version number that works with the currently release jazz_fingers 5.0.1...

```rb
[8] pry(main)> JazzFingers::VERSION
# => "5.0.1"
[9] pry(main)> Pry::VERSION
# => "0.10.0"

[5] pry(main)> JazzFingers
# => JazzFingers

[6] pry(main)> JazzFingers.configure
# NoMethodError: undefined method `configure' for JazzFingers:Module

[7] pry(main)> JazzFingers.setup!
# NoMethodError: undefined method `setup!' for JazzFingers:Module
```

Safe to ignore — `gem` was sourcing the modified version of 5.0.1 from the local cache instead of re-downloading it.

```rb
2.5.5 (jazz_fingers)[1] » Pry::VERSION
# "0.10.0"
2.5.5 (jazz_fingers)[2] » JazzFingers::VERSION
# "5.0.1"
2.5.5 (jazz_fingers)[3] » JazzFingers.setup!
# => true
```

h35514n replied:
Commit 857320f restores backwards-compatibility down to Pry ~> 0.10

85c5e86 refactors to isolate branching and move special casing to a different class — if that's too design-y or if you'd prefer a different class name, let me know.

```rb
# lib/jazz_fingers.rb L25-40 (85c5e865)

def prompt_class
  if Pry::VERSION >= "0.13.0"
    PromptPry013
  else
    PromptPry012
  end
end

def prompt
  @prompt ||=
    prompt_class.new(
      colored: config.colored_prompt,
      separator: config.prompt_separator,
      application_name: config.application_name
    )
end
```
<sup>[[source](https://github.com/jmromer/jazz_fingers/blob/85c5e865/lib/jazz_fingers.rb#L25-L40)]</sup>

h35514n replied:
Btw an alternative here would be to mix-in functionality by target pry version rather than inherit:

```rb
class Prompt
  if Pry::VERSION >= "0.13.0"
    include PromptPry013
  else
    include PromptPry012
  end
   # . . . 
end
```

h35514n replied:
Here's what that looks like: d2f71b3
Happy to keep either approach or change tack altogether.

Tested on pry 0.10.0 and 0.13.1:

```rb
2.5.5 (jazz_fingers)[1] » JazzFingers::VERSION
"5.0.1"

2.5.5 (jazz_fingers)[2] » Pry::VERSION
"0.10.0"
```

```rb

2.7.0 (jazz_fingers)[1] » JazzFingers::VERSION
"5.0.1"

2.7.0 (jazz_fingers)[2] » Pry::VERSION
"0.13.1"
```

plribeiro3000 commented on lib/jazz_fingers.rb:
```diff
@@ -28,7 +28,24 @@ def prompt
+        prompt.template(Pry.view_clip(context), pry, separator)
+      end
+
+      Pry::Prompt[:jazz_fingers]
```
What if we make this in the `Prompt` class instead?

There is a lot more going on on the `prompt` front then just `template(Pry.view_clip(context), pry, separator)`.

What if the `Prompt` class becomes responsible to add itself to `Pry::Prompt`?

h35514n replied:
> There is a lot more going on on the prompt front then just template(Pry.view_clip(context), pry, separator)

Can you disambiguate? I'm not sure what you mean by this.

> What if we make this in the Prompt class instead?

That's possible (that would be similar to 594223d). The current approach takes the `JazzFingers` module as the integration point with `Pry`, so the hard dependency on the `Pry` class is isolated there.

plribeiro3000 commented on lib/jazz_fingers/configuration.rb:
```diff
@@ -39,10 +39,10 @@ def awesome_print?
-      return "(#{underscore(@application_name)})" unless @application_name.nil?
-      return "(#{Rails.application.class.parent_name.underscore})" if defined?(Rails)
+      return underscore(@application_name) unless @application_name.nil?
+      return Rails.application.class.parent_name.underscore if defined?(Rails)
```
I like to have `()` around the project name to have a separated context.

Why the change?

h35514n replied:
There's no change in behavior / prompt formatting — the parentheses are just moved to the prompt class to reduce duplication:

```rb
# lib/jazz_fingers/prompt.rb L48-49 (2453e7e2)

  blue_text("(#{name})")
end
```
<sup>[[source](https://github.com/jmromer/jazz_fingers/blob/2453e7e2/lib/jazz_fingers/prompt.rb#L48-L49)]</sup>

plribeiro3000 replied:
Ohhh. Cool

plribeiro3000 commented on lib/jazz_fingers/prompt.rb:
```diff
@@ -32,44 +34,67 @@ def separator
+          module_name
+        end
+
+      blue_text("(#{name})")
```
Not sure i follow everything here but i guess once we figure out the other stuff i commented we can follow here.

h35514n replied:
Good call-out thanks — I added some documentation in 4b815c7

plribeiro3000 commented on lib/jazz_fingers/version.rb:
```diff
@@ -1,3 +1,3 @@
 module JazzFingers
-  VERSION = '5.0.1'.freeze
+  VERSION = "5.0.2".freeze
```
Please don't do this in the same commit/PR.

Commit:
Revert changelog and version number changes

Commit:
Add documentation for Prompt#context

Commit:
Relax pry dependency requirement

Commit:
Make backwards-compatible to Pry 0.10.0

- Moves Pry::Prompt-adding to JazzFingers::Prompt
- Conditionally sets the prompt using the previous approach

Commit:
Refactor to polymorphism

Commit:
Refactor to use mixins

h35514n commented on lib/jazz_fingers/prompt.rb:
```diff
@@ -28,48 +38,81 @@ def bold_text(text)
-      blue_text(@application_name)
+    def wait_separator
+      "*"
+    end
```
The naming here is updated for consistency with [Pry's](https://github.com/pry/pry/wiki/Customization-and-configuration#the-prompt) nomenclature:

> Pry can accept two prompt-types for every prompt - the 'main prompt' and the 'wait prompt'. The main prompt is always used for the first line of input; the wait prompt is used in multi-line input to indicate that the current expression is incomplete and more lines of input are required.

plribeiro3000 replied:
👍

h35514n commented on lib/jazz_fingers/prompt/pry_012.rb:
```diff
@@ -0,0 +1,24 @@
+        end
+      end
+
+      def wait_prompt
```
see: https://github.com/plribeiro3000/jazz_fingers/pull/23/files#r408413403

Commit:
Require mixin modules instead of autoloading

This resolves a failure when running tests with `bundle exec`:

```
% bundle exec rspec spec/workers/file.rb

Failure/Error: require File.expand_path("../../config/environment", __FILE__)

LoadError:
  cannot load such file -- jazz_fingers/prompt/pry_012

 ./config/initializers/jazz_fingers.rb:1:in `<top (required)>'
 ./config/environment.rb:8:in `<top (required)>'
 ./spec/rails_helper.rb:25:in `require'
 ./spec/rails_helper.rb:25:in `<top (required)>'
```

Commit:
Undo unneeded Hound changes, minimize diff

aried3r commented:
This looks great! Thanks for taking the time and putting so much effort into it. @plribeiro3000, WDYT?

plribeiro3000 commented on lib/jazz_fingers/prompt.rb:
```diff
@@ -1,5 +1,18 @@
+# frozen_string_literal: true
+
+require_relative "prompt/pry_012"
+require_relative "prompt/pry_013"
```
I rather use `autoload` here.

This way we won't load code that is not going to be used.

h35514n replied:
Using autoload for these modules creates an exception when running Pry in a Bundler context: https://github.com/plribeiro3000/jazz_fingers/pull/23/commits/ea6633bfea4ea68d7ba6855d57ba9b8a062ed718

I think these lines will only be triggered when `JazzFingers::Prompt` is autoloaded, but if there's another way you'd recommend I can test it out.

plribeiro3000 commented on lib/jazz_fingers/prompt.rb:
```diff
@@ -1,5 +1,18 @@
+    OBJECT_INSTANCE = /#<(.+)>/
+
+    if Pry::VERSION >= "0.13.0"
+      include Pry013
```
Maybe call this `Pry::Prompt::UnderVersion013` and `Pry::Prompt::AboveVersion012`?

Just a thought. I usually prefer names with more context. It makes easier for someone that is not aware of everything happening to debug it when/if needed.

WDYT?

h35514n replied:
Sounds good — updated in 4e0e5ff

plribeiro3000 commented on lib/jazz_fingers/prompt.rb:
```diff
@@ -28,48 +41,81 @@ def bold_text(text)
+    # the JazzFingers config. Examples: "(my_rails_app_name)", "(jazz_fingers)".
+    #
+    # When in the context of an object instance, use the abbreviated object
+    # path. Example: "(#<Pry::Prompt>)", "(#<...::ModuleName::ClassName>)"
```
COOOOOOOOOOL!

👏 👏 👏 👏

plribeiro3000 commented on lib/jazz_fingers/prompt.rb:
```diff
@@ -28,48 +41,81 @@ def bold_text(text)
-        "#{RUBY_VERSION} #{name}#{line_number(pry)} * #{spaces}"
-      end
+      ["#<...", tail.join("::"), ">"].join
     end
```
Now that i understand what this method does i'm a bit worried that its complexity might add some bugs.

I have some concerns here that if two difficult to address maybe we could try to test it out? Lets see if we can make this code a bit simplier.

PS: I know is harsh to test it. Thats one of the main reasons it does not have tests at all but i tried as much as possible to keep it simple so the necessity would not be too big.

plribeiro3000 commented on lib/jazz_fingers/prompt.rb:
```diff
@@ -28,48 +41,81 @@ def bold_text(text)
-        "(#{'../' * level}#{text})"
-      end
-    end
+      object_path_length = object_path_components.map(&:length).sum
```
Wouldn't the right length be based on object_path instead?
This result does not account fot `::`.

h35514n replied:
Accounting for the colons adds complexity to the subsequent iteration.
But: Only showing the root and leaf will be easier to maintain, so I'll pivot toward that instead. Updated in a7208be

plribeiro3000 commented on lib/jazz_fingers/prompt.rb:
```diff
@@ -28,48 +41,81 @@ def bold_text(text)
-      end
-    end
+      object_path_length = object_path_components.map(&:length).sum
+      return object_label if object_path_length < max_length
```
If we follow my last comment i believe this should be the first guard clause on this method.

plribeiro3000 commented on lib/jazz_fingers/prompt.rb:
```diff
@@ -28,48 +41,81 @@ def bold_text(text)
-        spaces = '  ' * level
-        "#{RUBY_VERSION} #{name}#{line_number(pry)} * #{spaces}"
-      end
+      ["#<...", tail.join("::"), ">"].join
```
Food for thought.

What if we just change this all together to always show the first and last module/class?

This way we won't loose the Root element neither the last one.

WDYT?

h35514n replied:
That works! simpler to maintain. Updated in a7208be

plribeiro3000 commented:
@jmromer I gotta say.

This has been a hell out of a good job understanding how `pry` works and how to make this change in a good way.

Than you a ton for your relentless efforts here. This feature will make it way much better.

❤️ ❤️ ❤️ ❤️ ❤️

plribeiro3000 commented on lib/jazz_fingers/prompt.rb:
```diff
@@ -28,48 +41,81 @@ def bold_text(text)
+    #
+    #   In:  #<NoPathJustASingleLongClassName>
+    #   Out: #<NoPathJustASingleLongClassName>
+    def abbreviated_context(object_label, max_length: 20)
```
Hmmmmm. Could we make `max_length` configurable?

Commit:
Abbreviate object path with root and leaf only.

Commit:
Name modules more descriptively

Commit:
Documentation update

plribeiro3000 commented:
Great work here @jmromer.

Sorry for the delays to review.

❤️ ❤️ ❤️ ❤️ ❤️ ❤️

plribeiro3000 commented:
Just released as `5.1.0.rc1`.

Will try it out in my projects and see if everything works ok.
